### DISCUSSION TITLE
'typeof(T).FullName' is never null

### DISCUSSION THREAD

**Discussion by sharwell** [dotnet/csharplang#3664] [Language Ideas]
**Version Used**: 3.7, targeting netcoreapp3.1

**Steps to Reproduce**:

```csharp
#nullable enable

class C
{
  string TypeName => typeof(C).FullName;
}
```

**Expected Behavior**:

No warning is reported for the expression returned from `TypeName`.

**Actual Behavior**:

> warning CS8603: Possible null reference return.

**Additional Information**:

While the `FullName` property can return `null` in some cases, it cannot return null for the syntax seen here. There is no way to annotate the property as having a non-null value conditioned on language-specific syntax producing the `Type` instance.

**Comment by tannergooding** [comment]
There are lots of APIs in .NET, especially in the reflection/metadata namespaces that will never return `null` except for some weird edge case (generally around dynamically created types).

Is this one really that special in comparison?

**Comment by jaredpar** [comment]
Moving to csharplang as this is really a language design topic. Essentially the question comes down to whether or not we want the language to override the annotations of the runtime library in specific contexts like `typeof`.

**Comment by Joe4evr** [comment]
Asked about this some time ago myself. Stephen gave a good explanation: https://github.com/dotnet/runtime/issues/31558#issuecomment-557932606

**Comment by jnm2** [comment]
This would be addressed by a proposal to make `typeof(C).FullName` a constant in that code sample (https://github.com/dotnet/csharplang/issues/8505).
